src/EventStore.Core.Tests/Services/Replication/LeaderReplication/when_replica_subscribes_with_epochs.cs

```diff
@@ -0,0 +1,343 @@
+		: with_replication_service_and_epoch_manager {
+		private readonly Guid _replicaId = Guid.NewGuid();
+		private TcpConnectionManager _replicaManager;
+		private List<Epoch> _replicaEpochs;
+
+		public override void When() {
+			EpochManager.WriteNewEpoch(0);
+			Writer.Write(CreateLogRecord(0, "test-stream", "test"), out _);
+			Writer.Write(CreateLogRecord(1, "test-stream", "test"), out _);
+			Writer.Write(CreateLogRecord(2, "test-stream", "test"), out _);
+			Writer.Write(CreateLogRecord(3, "test-stream", "test"), out _);
+			Writer.Write(CreateLogRecord(4, "test-stream", "test"), out _);
+			EpochManager.WriteNewEpoch(1);
+
+			var subscribePosition = Db.Config.WriterCheckpoint.ReadNonFlushed() + 1000;
+			_replicaEpochs = new List<Epoch> {
+				new Epoch(subscribePosition, 2, Guid.NewGuid()),
+			};
+			_replicaEpochs.AddRange(EpochManager.GetLastEpochs(10)
+				.Select(e => new Epoch(e.EpochPosition, e.EpochNumber, e.EpochId)).ToList());
+
+			AddSubscription(_replicaId, true, _replicaEpochs.ToArray(), subscribePosition, out _replicaManager);
+		}
+
+		[Test]
+		public void subscription_is_sent_replica_subscribed_message_for_leaders_writer_checkpoint() {
+			var subscribed = GetTcpSendsFor(_replicaManager).Select(x => x.Message)
+				.OfType<ReplicationMessage.ReplicaSubscribed>().ToArray();
+			Assert.AreEqual(1, subscribed.Length);
+			Assert.AreEqual(Db.Config.WriterCheckpoint.ReadNonFlushed(), subscribed[0].SubscriptionPosition);
+			Assert.AreEqual(_replicaId, subscribed[0].SubscriptionId);
+			Assert.AreEqual(LeaderId, subscribed[0].LeaderId);
+		}
+	}
+
+	public class when_replica_subscribes_with_additional_epoch_and_leader_has_epoch_after_common_epoch
+		: with_replication_service_and_epoch_manager {
+		private readonly Guid _replicaId = Guid.NewGuid();
+		private TcpConnectionManager _replicaManager;
+		private List<Epoch> _replicaEpochs;
+
+		public override void When() {
+			EpochManager.WriteNewEpoch(0);
+			Writer.Write(CreateLogRecord(0, "test-stream", "test"), out _);
+			Writer.Write(CreateLogRecord(1, "test-stream", "test"), out _);
+			Writer.Write(CreateLogRecord(2, "test-stream", "test"), out _);
+			EpochManager.WriteNewEpoch(1);
+			Writer.Write(CreateLogRecord(3, "test-stream", "test"), out _);
+			Writer.Write(CreateLogRecord(4, "test-stream", "test"), out _);
+			EpochManager.WriteNewEpoch(4);
+
+			var subscribePosition = Db.Config.WriterCheckpoint.ReadNonFlushed() + 1000;
+			_replicaEpochs = new List<Epoch> {
+				new Epoch(subscribePosition, 2, Guid.NewGuid()),
+			};
+			_replicaEpochs.AddRange(EpochManager.GetLastEpochs(10)
+				.Where(e => e.EpochNumber < 4)
+				.Select(e => new Epoch(e.EpochPosition, e.EpochNumber, e.EpochId)).ToList());
+
+			AddSubscription(_replicaId, true, _replicaEpochs.ToArray(), subscribePosition, out _replicaManager);
+		}
+
+		[Test]
+		public void subscription_is_sent_replica_subscribed_message_for_leaders_epoch_after_common_epoch() {
```

Comment:
i think this is specifically for the case where the leaders extra epoch is behind the replicas extra epoch